modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/matchers/source/SourceTransforms.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.logsdb.qa.matchers.source;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+class SourceTransforms {
+    /**
+     * This preprocessing step makes it easier to match the document using a unified structure.
+     * It performs following modifications:
+     * <ul>
+     * <li> Flattens all nested maps into top level map with full field path as key (e.g. "a.b.c.d") </li>
+     * <li> Transforms all field values to arrays of length >= 1 </li>
+     * </ul>
+     * <p>
+     * It also makes it possible to work with subobjects: false/auto settings.
+     *
+     * @return flattened map
+     */
+    public static Map<String, List<Object>> normalize(Map<String, Object> map) {
+        var flattened = new HashMap<String, List<Object>>();
+
+        descend(null, map, flattened);
+
+        return flattened;
+    }
+
+    public static <T> List<T> normalizeValues(List<T> values) {
+        if (values == null) {
+            return Collections.emptyList();
+        }
+
+        // Synthetic source modifications:
+        // * null values are not present
+        // * duplicates are removed
```

Comment:
We probably need to add an extension to randomization to inject duplicates.. 

In general, it'd be nice to have some values repeated randomly, esp for hostname.,

Reply:
That's a very good idea!